python/ray/data/datasource/file_meta_provider.py

```diff
@@ -365,3 +364,142 @@ def _handle_read_os_error(error: OSError, paths: Union[str, List[str]]) -> str:
         )
     else:
         raise error
+
+
+def _expand_paths(
+    paths: List[str],
+    filesystem: "pyarrow.fs.FileSystem",
+    partitioning: Optional[Partitioning],
+) -> Iterator[Tuple[str, int]]:
+    """Get the file sizes for all provided file paths."""
+    from pyarrow.fs import LocalFileSystem
+    from ray.data.datasource.file_based_datasource import (
+        FILE_SIZE_FETCH_PARALLELIZATION_THRESHOLD,
+        PATHS_PER_FILE_SIZE_FETCH_TASK,
+        _wrap_s3_serialization_workaround,
+        _unwrap_protocol,
+        _fetch_metadata_parallel,
+    )
+
+    # We break down our processing paths into a few key cases:
+    # 1. If len(paths) < threshold, fetch the file info for the individual files/paths
+    #    serially.
+    # 2. If all paths are contained under the same parent directory (or base directory,
+    #    if using partitioning), fetch all file infos at this prefix and filter to the
+    #    provided paths on the client; this should be a single file info request.
+    # 3. If more than threshold requests required, parallelize them via Ray tasks.
+
+    # 1. Small # of paths case.
+    if (
+        len(paths) < FILE_SIZE_FETCH_PARALLELIZATION_THRESHOLD
+        # Local file systems are very fast to hit.
+        or isinstance(filesystem, LocalFileSystem)
+    ):
+        for path in paths:
+            yield from _get_file_infos(path, filesystem)
+        return
+
+    # 2. Common path prefix case.
+    # Get longest common path of all paths.
+    common_path = os.path.commonpath(paths)
+    common_path = os.path.normpath(common_path)
+    # If parent directory (or base directory, if using partitioning) is common to all
+    # paths, fetch all file infos at that prefix and filter the response to the provided
+    # paths.
+    normalized_base_dir = _unwrap_protocol(partitioning.base_dir)
+    if (partitioning is not None and common_path == normalized_base_dir) or all(
+        str(pathlib.Path(path).parent) == common_path for path in paths
+    ):
+        path_to_size = {path: 0 for path in paths}
+        for path, file_size in _get_file_infos(common_path, filesystem):
+            if path in path_to_size:
+                path_to_size[path] = file_size
+        # Dictionaries are insertion-ordered, so this path + size pairs should be
+        # yielded in the order of the original paths arg.
+        yield from path_to_size.items()
```

Comment:
can we log a warning if any path in `path_to_size` not have size being set?

Reply:
Opted for an `assert` since this should never happen!